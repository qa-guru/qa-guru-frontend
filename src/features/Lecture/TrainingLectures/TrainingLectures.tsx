import React from "react";
import { CardActionArea, Grid, Paper, Typography } from "@mui/material";
import { Link } from "react-router-dom";
import styles from "./TrainingLectures.module.scss";
import { TrainingLecturesQuery } from "../../../generated/graphql";

interface ITrainingLectures {
  data: TrainingLecturesQuery;
  trainingId: string;
}

const TrainingLectures: React.FC<ITrainingLectures> = ({
  data,
  trainingId,
}) => {
  return (
    <Grid container spacing={2}>
      {data?.trainingLectures?.map((item, index) => {
        return (
          <Grid item xs={12} key={index}>
            <Link
              className={styles.link}
              to={`/training/${trainingId}/${item?.lecture?.id}`}
            >
              <CardActionArea>
                <Paper className={styles.paper}>
                  <Typography variant="h3">{item?.lecture?.subject}</Typography>
                  <Typography variant="subtitle2" mt="16px">
                    {item?.lecture?.description}
                  </Typography>
                  <Typography color="primary" variant="h4" align="right">
                    Продолжить
                  </Typography>
                </Paper>
              </CardActionArea>
            </Link>
          </Grid>
        );
      })}
    </Grid>
  );
};

export default TrainingLectures;
