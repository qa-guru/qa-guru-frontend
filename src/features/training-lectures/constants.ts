export const INDEX_OFFSET = 1;

export const ROUTES = {
  HOME: "/",
};
