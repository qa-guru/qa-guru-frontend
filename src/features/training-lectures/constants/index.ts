export { INDEX_OFFSET, ROUTES } from "./constants";
