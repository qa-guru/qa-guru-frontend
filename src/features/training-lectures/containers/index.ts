export { default } from "./training-lectures";
