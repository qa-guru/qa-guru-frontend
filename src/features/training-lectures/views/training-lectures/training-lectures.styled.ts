import { styled } from "@mui/system";
import { Link } from "react-router-dom";
import { Box, Grid, Paper, Stack, Typography } from "@mui/material";

export const StyledGridContainer = styled(Grid)(({ theme }) => ({
  gap: theme.spacing(3),
  margin: "30px 0",
}));

export const StyledLink = styled(Link)({
  textDecoration: "none",
});

export const StyledPaper = styled(Paper)(({ theme }) => ({
  padding: "15px",
  [theme.breakpoints.up("sm")]: {
    padding: "20px",
  },
}));

export const StyledWrapper = styled(Stack)(({ theme }) => ({
  gap: theme.spacing(0.5),
}));

export const StyledTypography = styled(Typography)(({ theme }) => ({
  minWidth: "40px",
  height: "40px",
  backgroundColor: theme.palette.app.primary,
  borderRadius: "50%",
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
  color: theme.palette.app.white,
}));

export const StyledStack = styled(Stack)(({ theme }) => ({
  marginTop: "15px",
  flexDirection: "row",
  gap: theme.spacing(1.5),
  alignItems: "center",
}));

export const StyledBox = styled(Box)({
  display: "flex",
  justifyContent: "flex-end",
});

export const StyledSubtitle = styled(Typography)(({ theme }) => ({
  color: theme.palette.app.primary,
  textTransform: "uppercase",
}));
