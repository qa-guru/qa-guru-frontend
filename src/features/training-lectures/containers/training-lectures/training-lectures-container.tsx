import { FC } from "react";
import { useParams } from "react-router-dom";

import { AppSpinner } from "shared/components/spinners";
import NoDataErrorMessage from "shared/components/no-data-error-message";
import {
  useTrainingLecturesQuery,
  useTrainingQuery,
} from "api/graphql/generated/graphql";
import { FETCH_POLICY } from "shared/constants";

import TrainingLectures from "../../views/training-lectures";

const TrainingLecturesContainer: FC = () => {
  const { trainingId } = useParams();

  const { data: dataTrainingLectures, loading: loadingTrainingLectures } =
    useTrainingLecturesQuery({
      variables: { id: trainingId! },
      fetchPolicy: FETCH_POLICY.CACHE_AND_NETWORK,
    });
  const { data: dataTraining, loading: loadingTraining } = useTrainingQuery({
    variables: { id: trainingId! },
    fetchPolicy: FETCH_POLICY.CACHE_AND_NETWORK,
  });

  if (loadingTrainingLectures || loadingTraining) return <AppSpinner />;
  if (!dataTrainingLectures || !dataTraining || !trainingId)
    return <NoDataErrorMessage />;

  return (
    <TrainingLectures
      trainingId={trainingId}
      dataTrainingLectures={dataTrainingLectures}
      dataTraining={dataTraining}
    />
  );
};

export default TrainingLecturesContainer;
