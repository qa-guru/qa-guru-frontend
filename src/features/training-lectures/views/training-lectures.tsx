import { FC } from "react";
import { CardActionArea, Container, Grid, Typography } from "@mui/material";
import { ButtonTrainingList } from "shared/components/buttons";

import { ITrainingLectures } from "./training-lectures.types";
import {
  StyledBox,
  StyledGridContainer,
  StyledLink,
  StyledPaper,
  StyledStack,
  StyledSubtitle,
  StyledTypography,
  StyledWrapper,
} from "./training-lectures.styled";
import { INDEX_OFFSET } from "../constants";

const TrainingLectures: FC<ITrainingLectures> = (props) => {
  const { dataTrainingLectures, trainingId, dataTraining } = props;
  const { trainingLectures } = dataTrainingLectures;
  const name = dataTraining?.training?.name;

  return (
    <Container>
      <ButtonTrainingList />
      <Typography variant="h2">{name}</Typography>
      <StyledGridContainer container>
        {trainingLectures?.map((item, index) => {
          const { id, subject, description } = item?.lecture || {};

          return (
            <Grid item xs={12} key={id}>
              <StyledLink to={`/training/${trainingId}/${id}`}>
                <CardActionArea>
                  <StyledPaper>
                    <Typography variant="h4">{subject}</Typography>
                    <StyledWrapper>
                      <StyledStack>
                        <StyledTypography variant="subtitle2">
                          {index + INDEX_OFFSET}
                        </StyledTypography>
                        <Typography variant="subtitle1">
                          {description}
                        </Typography>
                      </StyledStack>
                    </StyledWrapper>
                    <StyledBox>
                      <StyledSubtitle variant="body2">
                        Продолжить
                      </StyledSubtitle>
                    </StyledBox>
                  </StyledPaper>
                </CardActionArea>
              </StyledLink>
            </Grid>
          );
        })}
      </StyledGridContainer>
    </Container>
  );
};

export default TrainingLectures;
