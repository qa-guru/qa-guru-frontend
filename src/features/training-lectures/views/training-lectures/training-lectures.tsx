import { FC } from "react";
import {
  CardActionArea,
  Container,
  Grid,
  Typography,
  Chip,
  Box,
} from "@mui/material";
import {
  Lock as LockIcon,
  Schedule as ScheduleIcon,
  LockOpen as LockOpenIcon,
} from "@mui/icons-material";
import dayjs from "dayjs";

import CustomLink from "shared/components/custom-link";

import { ITrainingLectures } from "./training-lectures.types";
import {
  StyledBox,
  StyledGridContainer,
  StyledPaper,
  StyledStack,
  StyledSubtitle,
  StyledTypography,
  StyledWrapper,
} from "./training-lectures.styled";
import { INDEX_OFFSET } from "../../constants";

const TrainingLectures: FC<ITrainingLectures> = (props) => {
  const { dataTrainingLectures, trainingId, dataTraining } = props;
  const { trainingLectures } = dataTrainingLectures;
  const name = dataTraining?.training?.name;

  return (
    <Container>
      <Typography variant="h2">{name}</Typography>
      <StyledGridContainer container>
        {trainingLectures?.map((item, index) => {
          const { id, subject, description } = item?.lecture || {};
          const { locking, availableFrom, isAvailable } = item || {};

          const formatDate = (date: string | null | undefined) => {
            if (!date) return "";
            return dayjs(date).format("DD.MM.YYYY HH:mm");
          };

          const getStatusChip = () => {
            if (locking) {
              return (
                <Chip
                  icon={<LockIcon />}
                  label="Урок заблокирован"
                  color="error"
                  size="small"
                />
              );
            }
            if (!isAvailable && availableFrom) {
              return (
                <Chip
                  icon={<ScheduleIcon />}
                  label={`Доступен с ${formatDate(availableFrom)}`}
                  color="warning"
                  size="small"
                />
              );
            }
            return (
              <Chip
                icon={<LockOpenIcon />}
                label="Доступен"
                color="success"
                size="small"
              />
            );
          };

          const isLessonAccessible = !locking && isAvailable;

          return (
            <Grid item xs={12} key={id}>
              <CardActionArea disabled={!isLessonAccessible}>
                {isLessonAccessible ? (
                  <CustomLink path={`/training/${trainingId}/${id}`}>
                    <StyledPaper>
                      <Box
                        sx={{
                          display: "flex",
                          justifyContent: "space-between",
                          alignItems: "flex-start",
                          mb: 1,
                        }}
                      >
                        <Typography variant="h4">{subject}</Typography>
                        {getStatusChip()}
                      </Box>
                      <StyledWrapper>
                        {description?.map((desc, index) => (
                          <StyledStack key={index}>
                            <StyledTypography variant="subtitle2">
                              {index + INDEX_OFFSET}
                            </StyledTypography>
                            <Typography variant="subtitle1">{desc}</Typography>
                          </StyledStack>
                        ))}
                      </StyledWrapper>
                      <StyledBox>
                        <StyledSubtitle variant="body2">
                          Продолжить
                        </StyledSubtitle>
                      </StyledBox>
                    </StyledPaper>
                  </CustomLink>
                ) : (
                  <StyledPaper
                    sx={{ opacity: 0.6, cursor: "not-allowed !important" }}
                  >
                    <Box
                      sx={{
                        display: "flex",
                        justifyContent: "space-between",
                        alignItems: "flex-start",
                        mb: 1,
                      }}
                    >
                      <Typography variant="h4">{subject}</Typography>
                      {getStatusChip()}
                    </Box>
                    <StyledWrapper>
                      {description?.map((desc, index) => (
                        <StyledStack key={index}>
                          <StyledTypography variant="subtitle2">
                            {index + INDEX_OFFSET}
                          </StyledTypography>
                          <Typography variant="subtitle1">{desc}</Typography>
                        </StyledStack>
                      ))}
                    </StyledWrapper>
                    <StyledBox>
                      <StyledSubtitle variant="body2">
                        {locking ? "Урок недоступен" : "Скоро откроется"}
                      </StyledSubtitle>
                    </StyledBox>
                  </StyledPaper>
                )}
              </CardActionArea>
            </Grid>
          );
        })}
      </StyledGridContainer>
    </Container>
  );
};

export default TrainingLectures;
