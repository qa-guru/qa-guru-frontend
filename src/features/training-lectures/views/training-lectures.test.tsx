import { MemoryRouter } from "react-router-dom";
import TrainingLectures from "./training-lectures";
import { trainingLectures } from "../../../shared/mocks/training-lectures.mock";
import { training } from "../../../shared/mocks/training.mock";
import { render } from "../../../test/utils-test";

describe("TrainingLectures", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <TrainingLectures
          trainingId="1234"
          dataTrainingLectures={trainingLectures}
          dataTraining={training}
        />
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
