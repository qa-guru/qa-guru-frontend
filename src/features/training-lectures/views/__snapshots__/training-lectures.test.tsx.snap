// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrainingLectures > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1lexyjf-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-qn5tej-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-3ymmqj-MuiTypography-root"
      >
        К списку курсов
      </h6>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-h8ke29-MuiTypography-root"
    >
      string
    </h4>
    <div
      class="MuiGrid-root MuiGrid-container css-8pqjud-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <a
          class="css-ytumd6"
          href="/training/1234/string"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root css-1t3prmd-MuiButtonBase-root-MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ou7ve4-MuiPaper-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-97ruyc-MuiTypography-root"
              >
                string
              </h6>
              <div
                class="MuiStack-root css-1yc1p5z-MuiStack-root"
              >
                <div
                  class="MuiStack-root css-idd31a-MuiStack-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1p7jsv9-MuiTypography-root"
                  >
                    1
                  </h6>
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle1 css-v3tflm-MuiTypography-root"
                  >
                    string1
                  </h6>
                </div>
                <div
                  class="MuiStack-root css-idd31a-MuiStack-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1p7jsv9-MuiTypography-root"
                  >
                    2
                  </h6>
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle1 css-v3tflm-MuiTypography-root"
                  >
                    string1
                  </h6>
                </div>
              </div>
              <div
                class="MuiBox-root css-fw9znx"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2 css-1hhqpuv-MuiTypography-root"
                >
                  Продолжить
                </h6>
              </div>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
